### ISSUE TITLE
[GPU] Implement true GPU batch processing in GpuInferenceEngine::process_batch_gpu

### ISSUE THREAD

**Issue by EffortlessSteven** [EffortlessMetrics/BitNet-rs#298]
## Problem Description

The `GpuInferenceEngine::process_batch_gpu` function in `crates/bitnet-inference/src/gpu.rs` contains a stub implementation that processes inference requests sequentially instead of utilizing true GPU batch processing. This significantly impacts performance and GPU utilization for production inference workloads.

## Environment

- **Affected File**: `crates/bitnet-inference/src/gpu.rs`
- **Affected Function**: `GpuInferenceEngine::process_batch_gpu` (lines 426-451)
- **Impact**: GPU batch inference performance, throughput optimization
- **Features**: `--features gpu` (CUDA acceleration)
- **MSRV**: Rust 1.90.0

### Root Cause
The current implementation falls back to sequential processing with a comment indicating that "full batching would require more complex implementation":

```rust
pub fn process_batch_gpu(
    &self,
    requests: &[(Vec<u32>, GenerationConfig)],
) -> Result<Vec<Vec<u32>>> {
    if requests.len() == 1 {
        // Single request - no batching needed
        return Ok(vec![self.generate_tokens_gpu(&requests[0].0, &requests[0].1)?]);
    }
    
    // GPU batch processing
    let batch_start = Instant::now();
    
    // For now, process sequentially (full batching would require more complex implementation)
    let results: Result<Vec<_>> = requests
        .iter()
        .map(|(tokens, config)| self.generate_tokens_gpu(tokens, config))
        .collect();
    
    // Update batch processing metrics (incorrectly measuring sequential time)
    {
        let mut metrics = self.metrics.lock().unwrap();
        metrics.kernel_launch_overhead_ms = batch_start.elapsed().as_millis() as f64;
    }
    
    results
}
```

### Performance Impact
- **Throughput Loss**: Sequential processing eliminates GPU parallelism benefits
- **Memory Inefficiency**: Multiple separate forward passes instead of batched operations
- **Kernel Launch Overhead**: Increased CUDA kernel launch overhead per request
- **GPU Underutilization**: Poor hardware resource utilization compared to batched inference

### Misleading Metrics
The current implementation incorrectly records sequential processing time as `kernel_launch_overhead_ms`, providing misleading performance metrics for batch operations.

### Primary Solution: True GPU Batch Processing

Implement genuine GPU batch processing with the following approach:

#### 1. Batch Tensor Construction
- Combine input tokens from multiple requests into padded batch tensors
- Generate attention masks for variable-length sequences
- Handle padding and sequence length alignment

#### 2. Batched Forward Pass
- Execute single forward pass on GPU for entire batch
- Utilize tensor cores and mixed precision (FP16/BF16) when available
- Implement memory-efficient batching with configurable batch sizes

#### 3. Output Distribution
- Extract per-request results from batched logits
- Apply individual sampling configurations per request
- Maintain request ordering and metadata

### Implementation Architecture

```rust
pub fn process_batch_gpu(
    &self,
    requests: &[(Vec<u32>, GenerationConfig)],
) -> Result<Vec<Vec<u32>>> {
    if requests.is_empty() {
        return Ok(Vec::new());
    }

    if requests.len() == 1 {
        // Single request optimization - no batching overhead
        return Ok(vec![self.generate_tokens_gpu(&requests[0].0, &requests[0].1)?]);
    }

    let batch_start = Instant::now();
    
    // Step 1: Construct batch tensors with padding
    let (batched_input, attention_mask, request_metadata) = 
        self.prepare_batch_tensors(requests)?;
    
    // Step 2: Execute batched forward pass
    let batched_logits = self.forward_batch_gpu(&batched_input, &attention_mask)?;
    
    // Step 3: Extract and sample per-request results
    let results = self.extract_batch_results(&batched_logits, &request_metadata, requests)?;
    
    // Step 4: Update accurate batch processing metrics
    self.update_batch_metrics(batch_start.elapsed(), requests.len());
    
    Ok(results)
}
```

#### Batch Tensor Preparation
```rust
fn prepare_batch_tensors(
    &self,
    requests: &[(Vec<u32>, GenerationConfig)],
) -> Result<(BitNetTensor, BitNetTensor, BatchMetadata)> {
    let max_seq_len = requests.iter()
        .map(|(tokens, _)| tokens.len())
        .max()
        .unwrap_or(0);
    
    let batch_size = requests.len();
    let mut batched_tokens = Vec::with_capacity(batch_size * max_seq_len);
    let mut attention_mask = Vec::with_capacity(batch_size * max_seq_len);
    let mut metadata = BatchMetadata::new(batch_size);
    
    for (i, (tokens, _)) in requests.iter().enumerate() {
        // Add tokens with right-padding
        batched_tokens.extend_from_slice(tokens);
        attention_mask.extend(vec![1; tokens.len()]);
        
        // Pad to max length
        let padding_len = max_seq_len - tokens.len();
        batched_tokens.extend(vec![0; padding_len]); // Use pad token
        attention_mask.extend(vec![0; padding_len]);
        
        metadata.add_request(i, tokens.len(), max_seq_len);
    }
    
    let input_tensor = BitNetTensor::from_slice(
        &batched_tokens,
        &[batch_size, max_seq_len],
        &self.backend.device,
    )?;
    
    let mask_tensor = BitNetTensor::from_slice(
        &attention_mask,
        &[batch_size, max_seq_len],
        &self.backend.device,
    )?;
    
    Ok((input_tensor, mask_tensor, metadata))
}
```

#### Batched Forward Pass
```rust
fn forward_batch_gpu(
    &self,
    input: &BitNetTensor,
    attention_mask: &BitNetTensor,
) -> Result<BitNetTensor> {
    let compute_start = Instant::now();
    
    // Use model's forward pass with batch support
    let model = self.model.read().await;
    let logits = model.forward_with_attention_mask(input, attention_mask)?;
    
    // Update compute metrics
    {
        let mut metrics = self.metrics.lock().unwrap();
        metrics.compute_time_ms = compute_start.elapsed().as_millis() as f64;
    }
    
    Ok(logits)
}
```

#### Result Extraction
```rust
fn extract_batch_results(
    &self,
    batched_logits: &BitNetTensor,
    metadata: &BatchMetadata,
    requests: &[(Vec<u32>, GenerationConfig)],
) -> Result<Vec<Vec<u32>>> {
    let mut results = Vec::with_capacity(requests.len());
    
    for (i, (_, config)) in requests.iter().enumerate() {
        let request_logits = batched_logits.slice_batch(i)?;
        let generated_tokens = {
            let mut sampling = self.sampling.lock().unwrap();
            sampling.sample(&request_logits, &[], 0, config)?
        };
        results.push(generated_tokens);
    }
    
    Ok(results)
}
```

### Phase 1: Core Batch Processing (Priority: High)
- [ ] Implement `prepare_batch_tensors` method with padding and attention masks
- [ ] Create `BatchMetadata` struct to track request information
- [ ] Implement `forward_batch_gpu` with attention mask support
- [ ] Add `extract_batch_results` for per-request output extraction

### Phase 2: Optimization and Configuration (Priority: Medium)
- [ ] Add configurable maximum batch size (respect `GpuPerformanceConfig::max_batch_size`)
- [ ] Implement batch splitting for requests exceeding max batch size
- [ ] Add memory usage optimization for large batches
- [ ] Implement mixed precision batching (FP16/BF16) when supported

### Phase 3: Advanced Features (Priority: Low)
- [ ] Add dynamic batching based on sequence length similarity
- [ ] Implement batch-aware KV cache management
- [ ] Add batch scheduling and queuing optimizations
- [ ] Support for heterogeneous generation configurations within batches

### Phase 4: Performance Validation (Priority: High)
- [ ] Add comprehensive batch processing benchmarks
- [ ] Validate memory usage patterns with various batch sizes
- [ ] Cross-validate batch results against sequential processing
- [ ] Add performance regression tests

### Unit Tests
```rust
#[cfg(test)]
mod batch_processing_tests {
    use super::*;
    
    #[test]
    fn test_batch_tensor_preparation() {
        // Test padding, attention masks, metadata generation
    }
    
    #[test]
    fn test_batch_forward_pass() {
        // Test batched forward pass with mixed sequence lengths
    }
    
    #[test]
    fn test_result_extraction() {
        // Test per-request result extraction and ordering
    }
    
    #[test]
    fn test_batch_size_limits() {
        // Test behavior with max_batch_size configuration
    }
}
```

### Integration Tests
```rust
#[cfg(test)]
mod integration_tests {
    #[test]
    fn test_batch_vs_sequential_equivalence() {
        // Verify batch processing produces identical results to sequential
    }
    
    #[test]
    fn test_batch_performance_improvement() {
        // Validate performance gains from batch processing
    }
    
    #[test]
    fn test_memory_efficiency() {
        // Verify memory usage patterns and efficiency
    }
}
```

### Performance Benchmarks
```rust
#[bench]
fn bench_batch_processing_throughput(b: &mut Bencher) {
    // Benchmark tokens/second with various batch sizes
}

#[bench]
fn bench_batch_vs_sequential_latency(b: &mut Bencher) {
    // Compare latency characteristics
}
```

## Acceptance Criteria

- [ ] **Functional Requirements**:
  - `process_batch_gpu` performs true GPU batch processing instead of sequential iteration
  - Handles variable-length input sequences with proper padding and attention masks
  - Maintains identical output quality compared to sequential processing
  - Respects individual `GenerationConfig` settings per request
  
- [ ] **Performance Requirements**:
  - Achieves >2x throughput improvement for batch_size >= 4 compared to sequential processing
  - Reduces GPU kernel launch overhead through batched operations
  - Maintains memory usage within configured limits
  
- [ ] **Compatibility Requirements**:
  - Integrates with existing `GpuPerformanceConfig::max_batch_size` settings
  - Works with both mixed precision and standard precision modes
  - Compatible with KV cache system and memory management
  
- [ ] **Quality Requirements**:
  - Includes comprehensive unit and integration tests
  - Provides accurate performance metrics (no more misleading measurements)
  - Follows BitNet.rs coding standards and documentation requirements

### Implementation Risks
- **Memory Pressure**: Large batches may exceed GPU memory limits
  - *Mitigation*: Implement batch size validation and splitting
- **Numerical Precision**: Batched operations may introduce floating-point differences
  - *Mitigation*: Cross-validation testing and deterministic mode support
- **Complexity**: Batched attention and KV cache management adds implementation complexity
  - *Mitigation*: Incremental implementation with thorough testing

### Performance Risks
- **Padding Overhead**: Variable sequence lengths require padding, potentially wasting computation
  - *Mitigation*: Smart batching by sequence length similarity
- **Memory Bandwidth**: Large batch tensors may be memory bandwidth bound
  - *Mitigation*: Profiling and optimization of tensor layouts

## Related Issues

This issue addresses GPU batch processing and is related to:
- Backend Implementation Issues (tokenization, forward passes)
- GPU Memory Management Issues (efficient memory allocation for batches) 
- Performance Optimization Issues (throughput improvements)
- Validation and Testing Issues (cross-validation of batch vs sequential results)

## Definition of Done

This issue will be considered complete when:
1. `GpuInferenceEngine::process_batch_gpu` implements true GPU batch processing
2. All unit and integration tests pass with >95% code coverage
3. Performance benchmarks show measurable throughput improvements
4. Cross-validation confirms output equivalence with sequential processing
5. Documentation and examples are updated to reflect new capabilities
6. Performance regression tests are in place to prevent future degradation

---

**Labels**: `enhancement`, `gpu`, `performance`, `inference-engine`
**Priority**: High
**Estimated Effort**: 3-5 days
